libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3956,18 +3973,47 @@ bool TypeChecker::visit(Literal const& _literal)
 		[&](Literal::SubDenomination) {},
 	}, _literal.suffix());
 
+	FunctionDefinition const* literalSuffixFunction = std::visit(util::GenericVisitor{
+		[&](ASTPointer<Identifier> const& _identifier) {
+			return dynamic_cast<FunctionDefinition const*>(_identifier->annotation().referencedDeclaration);
+		},
+		[&](ASTPointer<MemberAccess> const& _memberAccess) {
+			return dynamic_cast<FunctionDefinition const*>(_memberAccess->annotation().referencedDeclaration);
+		},
+		[&](Literal::SubDenomination) -> FunctionDefinition const* { return nullptr; },
+	}, _literal.suffix());
```

Comment:
I first tried using method `suffixFunction` from `Literal` class, but it has an `assert` that fails when the `referencedDeclaration` is not a `FunctionDefinition`, as in `syntaxTests/literalSuffixes/shadowed_suffix.sol`.

Reply:
You should not try to access the function here because `visit(Literal)` runs before `visit(Identifier)`, which does overload resolution. Anything that needs access to the function must go into `endVisit(Literal)` instaed.

But I'm wondering if we still need this error in the first place. It was useful when messages in `endVisit(Literal)` were very generic. Now that you made them more detailed, they already cover this.

Please check if removing this whole `if` leaves any error case undetected. If it does not, we can remove this whole check.

Reply:
We cannot remove the `if (!mantissa || !exponent)` because later we rely on that in https://github.com/ethereum/solidity/blob/65bdf49c53d76a0b8049a660572aab4d1b942e7e/libsolidity/codegen/ExpressionCompiler.cpp#L736

Reply:
> You should not try to access the function here because visit(Literal) runs before visit(Identifier), which does overload resolution. Anything that needs access to the function must go into endVisit(Literal) instaed.

Things changed a little bit since this suggestion and I am not sure where I should move this error check... maybe create `endVisit(FunctionCall)` ?

Reply:
Ok, you're right, we still need it. It should be in `typeCheckSuffixFunctionCall()` though. I moved it next to the other check.